hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpBlockInfoSubcommand.java

```diff
@@ -0,0 +1,150 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.services.cli.signedstate;
+
+import static com.hedera.node.app.service.mono.state.submerkle.RichInstant.fromJava;
+import static com.hedera.services.cli.utils.ThingsToStrings.quoteForCsv;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.node.app.service.mono.state.merkle.MerkleNetworkContext;
+import com.hedera.node.app.service.mono.state.submerkle.RichInstant;
+import com.hedera.services.cli.signedstate.DumpTokensSubcommand.FieldBuilder;
+import com.hedera.services.cli.utils.ThingsToStrings;
+import com.hedera.services.cli.utils.Writer;
+import com.swirlds.base.utility.Pair;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/** Dump all block info from a signed state file to a text file in a deterministic order  */
+public class DumpBlockInfoSubcommand {
+
+    static void doit(@NonNull final SignedStateHolder state, @NonNull final Path blockInfoPath) {
+        new DumpBlockInfoSubcommand(state, blockInfoPath).doit();
+    }
+
+    @NonNull
+    final SignedStateHolder state;
+
+    @NonNull
+    final Path blockInfoPath;
+
+    DumpBlockInfoSubcommand(@NonNull final SignedStateHolder state, @NonNull final Path blockInfoPath) {
+        requireNonNull(state, "state");
+        requireNonNull(blockInfoPath, "blockInfoPath");
+
+        this.state = state;
+        this.blockInfoPath = blockInfoPath;
+    }
+
+    void doit() {
+        final var networkContext = state.getNetworkContext();
+        System.out.printf("=== block info ===%n");
+
+        final var blockInfo = new BlockInfo(networkContext);
+
+        int reportSize;
+        try (@NonNull final var writer = new Writer(blockInfoPath)) {
+            reportOnBlockInfo(writer, blockInfo);
+            reportSize = writer.getSize();
+        }
+
+        final var runningHashLeaf = state.getRunningHashLeaf();
+
+        System.out.printf("=== block info report is %d bytes%n", reportSize);
+    }
+
+    @SuppressWarnings(
+            "java:S6218") // "Equals/hashcode method should be overridden in records containing array fields" - this
+    record BlockInfo(
+            long lastBlockNumber,
+            @NonNull String blockHashes,
+            @Nullable RichInstant consTimeOfLastHandledTxn,
+            boolean migrationRecordsStreamed,
+            @Nullable RichInstant firstConsTimeOfCurrentBlock) {
+        BlockInfo(@NonNull final MerkleNetworkContext networkContext) {
+            this(
+                    networkContext.getAlignmentBlockNo(),
+                    networkContext.stringifiedBlockHashes(),
+                    fromJava(networkContext.consensusTimeOfLastHandledTxn()),
```

Comment:
I had to look up `fromJava` ... wasn't obvious to me where it came from ... perhaps it shouldn't be a static import and instead fully qualify it `RichInstant.fromJava(...)`?